server/src/shared/app/base-controller.ts

```diff
@@ -1,23 +1,42 @@
 import * as express from 'express'
-import { ValidationError } from 'joi'
-import { Result } from '../core/result'
+import { BaseUseCase } from './use-case'
+import Joi, { ValidationError } from 'joi'
+
+type extractDTO<UseCase> = UseCase extends BaseUseCase<infer T, any> ? T : never;
+
+export abstract class BaseController<UseCase extends BaseUseCase<any, any>> {
+  constructor(
+    protected readonly useCase: UseCase,
+    protected readonly schema: Joi.ObjectSchema<extractDTO<UseCase>>
+  ) {}
+
+  protected validate(body: unknown, onErr: (err: ValidationError) => void): body is extractDTO<UseCase> {
```

Comment:
Is there a way to use the `Result<T,E>` type instead of the `onErr` callback but keeping the `body is X` typeguard? We shouldn't mix error propagation styles in our codebase. We should make sure we stay consistent between

1. `throw Error()` and `try/catch`
2. using `Result<T,E>`
3. error handler with callbacks

If we can't somehow use `Result<T,E>` *and* narrow the type of `body with the type guard, then I think `onErr` or `try/catch` is ok 👍

Reply:
https://github.com/maghis/types-joi
https://github.com/hjkcai/typesafe-joi

I don't trust either. I'm going to switch to try/catch

Reply:
i think having a callback error is better than throwing one and having to catch it with `try/catch`. At least having a callback handler is still encoded in the type system.

Reply:
Yeah after implementing try/catch I realized it was unusable because typescript still expects something to happen if validate returns false. It can't return false, since it would just throw instead, so then we have awkward code written to handle an impossible case in the control flow